Docs network best 2.5

gundalow opened the pull request:
##### SUMMARY
Document best practices for 2.5 Networking, which details:

* Correct way for credentials to be specified
* How to use ``network_cli``
* How to use ``netconf``
* What become mode is, and how to use it

This is best practice. It is assumed readers will have good working Ansible and Ansible Network knowledge by this point


In future PRs the following will be added
* Link to platform specific guides (eAPI, NETCONF, NXAPI, etc)
* A separate guide for Ansible 21.->2.4 will be created

##### ISSUE TYPE
 - Docs Pull Request

gundalow committed:
Best practices for Ansible 2.5 networking

* How to use network_cli
* How to pass credentials

gundalow committed:
Proxy

IPvSean commented:
Thoughts (shouldn't stop the PR, but maybe we can add to another one)

- Can get a list of all available `ansible_network_os` . i.e. ios, nxos, eos, etc where we talk about the `ansible_network_os`
- This might sound dumb... but maybe we can make a picture and put it up there?  I can try drafting up something with you
- Before we dive into group_vars, etc maybe we could have a folder structure for the demo, like literally just do the `tree` command and show that output so they can see our layout?
- Are we using a ansible.cfg file for this?  inventory is not a normal name so I imagine you set this?  Maybe we can take your literal example and put it on our github.com/network-automation github repo !!
- One or more network devices that are compatible with Ansible.  Maybe we link to this-> https://access.redhat.com/articles/3185021
- At the bottom can we link to the network-automation github repo? github.com/network-automation
- The only thing "confusing" to me at all, is the "Collect data" part.  Is that the name of the example playbook or is this where you talk about gathering facts and explaining that?  I feel like you should call the playbook like backup.yml or something and then we put it on github.  I am just confused at the section title Collect data... its just a bit confusing if thats the name of the playbook, a sub-section or what.... maybe its just the way its rendering in markdown versus the docs page

gundalow commented:
> Can get a list of all available `ansible_network_os` . i.e. ios, nxos, eos, etc where we talk about the `ansible_network_os`

Good point, I think I'll add this to intro_network

> This might sound dumb... but maybe we can make a picture and put it up there?  I can try drafting up something with you

Picture of what?

> Before we dive into group_vars, etc maybe we could have a folder structure for the demo, like literally just do the `tree` command and show that output so they can see our layout?

`tree` is good idea

> Are we using a ansible.cfg file for this?  inventory is not a normal name so I imagine you set this? 

No ansible.cfg, I'm just doing
`ansible-playbook -i inventory fetch-facts.yml` where `-i` means "use this file for inventory

> Maybe we can take your literal example and put it on our github.com/network-automation github repo !!

Maybe, will think more on this

>One or more network devices that are compatible with Ansible.  Maybe we link to this-> https://access.redhat.com/articles/3185021

Not seem that page before, though that only lists what has "commercial supported", `dellos*` (for example) supports network_cli, though isn't covered by the "Network Addon"

> At the bottom can we link to the network-automation github repo? github.com/network-automation

Sure, do we have a few lines summary you could provide to explain what is is?

> The only thing "confusing" to me at all, is the "Collect data" part.  Is that the name of the example playbook or is this where you talk about gathering facts and explaining that?  I feel like you should call the playbook like backup.yml or something and then we put it on github.  I am just confused at the section title Collect data... its just a bit confusing if thats the name of the playbook, a sub-section or what.... maybe its just the way its rendering in markdown versus the docs page

I'll have a think about this

gundalow committed:
rename facts-demo.yml + tree

Qalthos commented on docs/docsite/rst/network_best_practices_2.5.rst:
```diff
@@ -0,0 +1,390 @@
+           var: hostvars['vyos01.example.net']
+
+       - name: Write facts to disk using a template
+copy:
```
Looks like some indentation got lost?

Qalthos replied:
Ah, ansibot saw it, too

IPvSean commented:
>Picture of what?

Was thinking of super simple diagram since we are not talking about anything but backing up configurations here... but something just to break the wall of text up.![picture](https://i.imgur.com/2Fk63Uu.png)

>Sure, do we have a few lines summary you could provide to explain what is is?

The Network-Automation GitHub project is a open community for all things network automation.  Here there will be various examples, demonstrations and Github repos dedicated exclusively for automation networking platforms.  To join simply email ansible-network@redhat.com and we will get you added right away!

>Not seem that page before, though that only lists what has "commercial supported", dellos* (for example) supports network_cli, though isn't covered by the "Network Addon"

Yeah I don't know if we have a great list of "everything yet" so maybe just word it like, List of all Red Hat supported networking modules until we get the community supported list as well (or link to the module index page for networking)

gundalow committed:
code-block formatting

IPvSean commented:
branch base was moved/deprecated, adding new Pull Request

gundalow committed:
Move platform specific notes to new page (PR pending)

gundalow commented:
* Removed platform specific notes, this will be documented on pages linked from `intro_network.rst`
* Corrected link to `intro_network`

@IPvSean I'll put a diagram on `intro_network.rst`

ismc commented:
It seems confusing to use ansible_network_os at times and the group (when grouped by os) at others, especially when you are setting ansible_network_os in group_vars.

privateip commented on docs/docsite/rst/network_best_practices_2.5.rst:
```diff
@@ -0,0 +1,372 @@
+
+  The ansible-connection setting tells Ansible how it should connect to a remote device. When working with Ansible Networking, setting this to ``ansible_connection: network_cli`` informs Ansible that the remote node is a network device with a limited execution environment. Without this setting, Ansible would attempt to use ssh to connect to the remote and execute the Python script on the network  ...
+:ansible_network_os:
+  Informs Ansible which Network platform this hosts corresponds to. This is required when using ``network_cli``
```
required for both `network_cli` and `netconf`

privateip commented on docs/docsite/rst/network_best_practices_2.5.rst:
```diff
@@ -0,0 +1,372 @@
+       #
+       - name: Gather facts (eos)
+         eos_facts:
+         when: "'eos' in group_names"
```
`when: ansible_network_os == 'eos'`

privateip commented on docs/docsite/rst/network_best_practices_2.5.rst:
```diff
@@ -0,0 +1,372 @@
+
+       - name: Gather facts (vyos)
+         vyos_facts:
+         when: "'vyos' in group_names"
```
as above (except vyos)

privateip commented on docs/docsite/rst/network_best_practices_2.5.rst:
```diff
@@ -0,0 +1,372 @@
+
+       - name: Gather facts (ops)
+         ios_facts:
+         when: "'ios' in group_names"
```
as above (except ios)

privateip commented on docs/docsite/rst/network_best_practices_2.5.rst:
```diff
@@ -0,0 +1,372 @@
+         eos_config:
+           backup: yes
+         register: backup_eos_location
+         when: "'eos' in group_names"
```
`when:` as above

dharmabumstead commented:
Doc review in process. Please don’t merge.

gundalow committed:
use ansible_network_os rather than group_names

gundalow commented on docs/docsite/rst/network_best_practices_2.5.rst:
```diff
@@ -0,0 +1,372 @@
+           content: |
+             #jinja2: lstrip_blocks: True
+             EOS device info:
+               {% for host in groups['eos'] %}
```
Is there a way of rewriting this using `ansible_network_os`?

gundalow replied:
This is apparently a fine use of groups still.

dharmabumstead commented on docs/docsite/rst/network_best_practices_2.5.rst:
```diff
@@ -0,0 +1,372 @@
+Connection & Credentials (group_vars)
+-------------------------------------
+
+As Ansible is a flexible tool there are a number of ways of specifying connection information & credentials. We believe however that the best was is to use ``group_vars``.
```
Please revise to: Because Ansible is a flexible tool, there are a number of ways to specify connection information and credentials. We recommend using ``group_vars``.

dharmabumstead commented on docs/docsite/rst/network_best_practices_2.5.rst:
```diff
@@ -0,0 +1,372 @@
+
+.. FIXME FUTURE Gundalow - Link to network auth & proxy page (to be written)
+
+.. warning:: Never store passwords in plain text
```
Period

dharmabumstead commented on docs/docsite/rst/network_best_practices_2.5.rst:
```diff
@@ -0,0 +1,372 @@
+
+.. warning:: Never store passwords in plain text
+
+The "Vault" feature of Ansible allows keeping sensitive data such as passwords or keys in encrypted files, rather than as plain text in your playbooks or roles. These vault files can then be distributed or placed in source control. See :doc:`playbooks_vault` for more information.
```
"allows you to keep" rather than "keeping sensitive data"

dharmabumstead commented on docs/docsite/rst/network_best_practices_2.5.rst:
```diff
@@ -0,0 +1,372 @@
+
+:ansible_connection:
+
+  The ansible-connection setting tells Ansible how it should connect to a remote device. When working with Ansible Networking, setting this to ``ansible_connection: network_cli`` informs Ansible that the remote node is a network device with a limited execution environment. Without this setting, Ansible would attempt to use ssh to connect to the remote and execute the Python script on the network  ...
```
Suggest revising to: Ansible uses the ``ansible-connection`` setting to determine how to connect to a remote device. When working with Ansible Networking, set this to netweork_cli so Ansible treats the remote node as a network device with a limited execution environment.

dharmabumstead commented on docs/docsite/rst/network_best_practices_2.5.rst:
```diff
@@ -0,0 +1,372 @@
+
+The :ref:`eos_config <eos_config>` and :ref:`vyos_config <vyos_config>` modules have a ``backup:`` option that when set will cause the module to create a full backup of the current ``running-config`` from the remote device before any changes are made. The backup file is written to the ``backup`` folder in the playbook root directory. If the directory does not exist, it is created.
+
+To demonstrate how we can move the backup file to a different location we ``register`` the result and use the ``backup_path`` return value as source location to move the file into ``/tmp/backups/`` directory which we have created.
```
Hard to parse. If I'm understanding it correctly, maybe try: "To demonstrate how we can move the backup file to a different location, we ``register`` the result and move the file to the path stored in ``backup_path``."

dharmabumstead commented on docs/docsite/rst/network_best_practices_2.5.rst:
```diff
@@ -0,0 +1,372 @@
+Troubleshooting
+===============
+
+If you receive an connection error please double check the inventory and Playbook for typos or missing lines, if the issue still occurs follow the debug steps in :doc:`network_debug_troubleshooting`.
```
Period instead of comma.

dharmabumstead reviewed (changes requested):
A few suggestions. Thanks @gundalow!

gundalow committed:
Scott's comments

gundalow commented:
@dharmabumstead Updated based on your feedback, thanks

Ready for final review

dharmabumstead reviewed (approved):
Looks good. Thanks @gundalow!

gundalow committed:
Update network_best_practices_2.5.rst

gundalow commented:
rebuild_merge
